code/controllers/subsystem/pathfinder.dm

```diff
@@ -38,10 +43,165 @@ SUBSYSTEM_DEF(pathfinder)
+		var/list/datum/path_map/owned_maps = currentmaps[source]
+		for(var/datum/path_map/map as anything in owned_maps)
+			if(map.creation_time < oldest_time && !map.building)
+				source_to_maps[source] -= map
+			owned_maps.len--
+			if(MC_TICK_CHECK)
+				return
+		if(!length(source_to_maps[source]))
+			source_to_maps -= source
+
+		currentmaps.len--
+
 /// Initiates a pathfind. Returns true if we're good, FALSE if something's failed
-/datum/controller/subsystem/pathfinder/proc/pathfind(atom/movable/caller, atom/end, max_distance = 30, mintargetdist, id=null, simulated_only = TRUE, turf/exclude, skip_first=TRUE, diagonal_safety=TRUE, datum/callback/on_finish)
-	var/datum/pathfind/path = new(caller, end, id, max_distance, mintargetdist, simulated_only, exclude, skip_first, diagonal_safety, on_finish)
+/datum/controller/subsystem/pathfinder/proc/pathfind(atom/movable/caller, atom/end, max_distance = 30, mintargetdist, access = list(), simulated_only = TRUE, turf/exclude, skip_first = TRUE, diagonal_handling = DIAGONAL_REMOVE_CLUNKY, list/datum/callback/on_finish)
+	var/datum/pathfind/jps/path = new()
+	path.setup(caller, access, max_distance, simulated_only, exclude, on_finish, end, mintargetdist, skip_first, diagonal_handling)
+	if(path.start())
+		active_pathing += path
+		return TRUE
+	return FALSE
+
+/// Initiates a swarmed pathfind. Returns true if we're good, FALSE if something's failed
+/// If a valid pathmap exists for the TARGET turf we'll use that, otherwise we have to build a new one
+/datum/controller/subsystem/pathfinder/proc/swarmed_pathfind(atom/movable/caller, atom/end, max_distance = 30, mintargetdist = 0, age = MAP_REUSE_INSTANT, access = list(), simulated_only = TRUE, turf/exclude, skip_first = TRUE, list/datum/callback/on_finish)
+	var/turf/target = get_turf(end)
+	var/datum/can_pass_info/pass_info = new(caller, access)
+	// If there's a map we can use already, use it
+	var/datum/path_map/valid_map = get_valid_map(pass_info, target, simulated_only, exclude, age, include_building = TRUE)
+	if(valid_map && valid_map.expand(max_distance))
+		path_map_passalong(on_finish, get_turf(caller), mintargetdist, skip_first, valid_map)
+		return TRUE
+
+	// Otherwise we're gonna make a new one, and turn it into a path for the callbacks passed into us
+	var/list/datum/callback/pass_in = list()
+	pass_in += CALLBACK(GLOBAL_PROC, /proc/path_map_passalong, on_finish, get_turf(caller), mintargetdist, skip_first)
+	// And to allow subsequent calls to reuse the same map, we'll put a placeholder in the cache, and fill it up when the pathing finishes
+	var/datum/path_map/empty = new()
+	empty.pass_info = new(caller, access)
+	empty.start = target
+	empty.pass_space = simulated_only
+	empty.avoid = exclude
+	empty.building = TRUE
+	path_map_cache(target, empty)
+	pass_in += CALLBACK(src, PROC_REF(path_map_fill), target, empty)
+	if(!SSpathfinder.can_pass_build_map(pass_info, target, max_distance, simulated_only, exclude, pass_in))
+		return null
```

Comment:
i'm not fond of having the `null`/`TRUE` dichotomy here, can this really not be `FALSE`?

Reply:
yeah no I just fucked it up see doc comment